diff --git a/mlir/lib/Bindings/Python/IRAttributes.cpp b/mlir/lib/Bindings/Python/IRAttributes.cpp
index 3685ff0d602..96308911dcc 100644
--- a/mlir/lib/Bindings/Python/IRAttributes.cpp
+++ b/mlir/lib/Bindings/Python/IRAttributes.cpp
@@ -975,8 +975,8 @@ nb::int_ PyDenseIntElementsAttribute::dunderGetItem(intptr_t pos) const {
   // Index type can also appear as a DenseIntElementsAttr and therefore can be
   // casted to integer.
   assert(mlirTypeIsAInteger(type) ||
-         mlirTypeIsAIndex(type) && "expected integer/index element type in "
-                                   "dense int elements attribute");
+         (mlirTypeIsAIndex(type) && "expected integer/index element type in "
+                                   "dense int elements attribute"));
   // Dispatch element extraction to an appropriate C function based on the
   // elemental type of the attribute. nb::int_ is implicitly
   // constructible from any C++ integral type and handles bitwidth correctly.
